File: checks/python/src/section508/checks/name_role_value.py
import re

from section508.dom import SPACE, attr, class_string, closest, find_id, has_attr, select, tag_name

STATE_FOR = {
    "tab": "aria-selected",
    "checkbox": "aria-checked",
    "radio": "aria-checked",
    "switch": "aria-checked",
    "menuitemcheckbox": "aria-checked",
    "slider": "aria-valuenow",
}

CLASS_WIDGET = re.compile(r"\b(tab|tabs|switch|accordion|menu-item|menuitem)\b", re.I)


def check(root):
    if not root:
        return []

    widgets = find_widgets(root)
    if not widgets:
        return []

    for widget in widgets:
        role = (attr(widget, "role") or "").lower()
        if not role:
            return [f"“{hint(widget)}” has no role."]
        if not has_name(widget, root):
            return [f"“{hint(widget)}” has no name."]
        state = STATE_FOR.get(role)
        if state and not has_attr(widget, state):
            return [f"“{hint(widget)}” has no {state}."]
    return []


def find_widgets(root):
    found = []
    for element in select(root, "div, span, button, a, [role]"):
        if closest(element, "button, a[href], label") and tag_name(element) not in {"button", "a"}:
            continue
        role = (attr(element, "role") or "").lower()
        if role in STATE_FOR or role == "menuitem":
            found.append(element)
            continue
        if CLASS_WIDGET.search(class_string(element)) and not is_native(element):
            found.append(element)
            continue
        if not is_native(element) and (has_attr(element, "aria-selected") or has_attr(element, "aria-checked")):
            found.append(element)
    return found


def is_native(element) -> bool:
    return tag_name(element) in {"button", "a", "input", "select", "textarea", "summary"}


def has_name(element, root) -> bool:
    aria = SPACE.sub(" ", attr(element, "aria-label") or "").strip()
    if aria:
        return True
    ids = [ref for ref in (attr(element, "aria-labelledby") or "").split() if ref]
    if any((find_id(root, ref).get_text() if find_id(root, ref) is not None else "").strip() for ref in ids):
        return True
    return SPACE.sub(" ", element.get_text()).strip() != ""


def hint(element) -> str:
    return (
        SPACE.sub(" ", element.get_text()).strip()
        or attr(element, "aria-label")
        or class_string(element)
        or tag_name(element)
    )

File: checks/ruby/lib/section508/checks/name_role_value.rb
# frozen_string_literal: true

module Section508
  module Checks
    # Name, role, and value for custom widgets only.
    class NameRoleValue
      STATE_FOR = {
        "tab" => "aria-selected",
        "checkbox" => "aria-checked",
        "radio" => "aria-checked",
        "switch" => "aria-checked",
        "menuitemcheckbox" => "aria-checked",
        "slider" => "aria-valuenow"
      }.freeze

      CLASS_WIDGET = /\b(tab|tabs|switch|accordion|menu-item|menuitem)\b/i
      NATIVE = %w[BUTTON A INPUT SELECT TEXTAREA SUMMARY].freeze

      def self.call(root)
        return [] unless root&.respond_to?(:css)

        widgets = find_widgets(root)
        return [] if widgets.empty?

        widgets.each do |widget|
          role = widget["role"].to_s.downcase
          return ["“#{hint(widget)}” has no role."] if role.empty?
          return ["“#{hint(widget)}” has no name."] unless named?(widget, root)

          state = STATE_FOR[role]
          return ["“#{hint(widget)}” has no #{state}."] if state && !Dom.has_attr?(widget, state)
        end
        []
      end

      def self.find_widgets(root)
        Dom.elements(root, "div, span, button, a, [role]").select do |element|
          if Dom.closest(element, "button, a[href], label") && !%w[BUTTON A].include?(element.name.upcase)
            next false
          end

          role = element["role"].to_s.downcase
          next true if STATE_FOR.key?(role) || role == "menuitem"
          next true if CLASS_WIDGET.match?(element["class"].to_s) && !native?(element)
          next true if !native?(element) && (Dom.has_attr?(element, "aria-selected") || Dom.has_attr?(element, "aria-checked"))

          false
        end
      end

      def self.native?(element)
        NATIVE.include?(element.name.upcase)
      end

      def self.named?(element, root)
        aria = element["aria-label"].to_s.gsub(/\s+/, " ").strip
        return true unless aria.empty?

        ids = element["aria-labelledby"].to_s.split(/\s+/).reject(&:empty?)
        return true if ids.any? { |id| (node = Dom.by_id(root, id)) && !Dom.text(node).strip.empty? }

        !Dom.norm(element).empty?
      end

      def self.hint(element)
        Dom.norm(element).then { |text| text.empty? ? nil : text } ||
          element["aria-label"] ||
          element["class"] ||
          element.name.downcase
      end
    end
  end
end
